plutus-ledger-api/src/PlutusLedgerApi/Common/SerialisedScript.hs

```diff
@@ -55,6 +60,38 @@ we only perform it in V2 and above.
 -- | Scripts to the ledger are serialised bytestrings.
 type SerialisedScript = ShortByteString
 
+{-| Note [Using Flat for serialsing/deserialising Script]
+`plutus-ledger` uses CBOR for data serialisation and `plutus-core` uses Flat. The
+choice to use Flat was made to have a more efficient (most wins are in uncompressed
+size) data serialisation format and use less space on-chain.
+
+To make `plutus-ledger` work with scripts serialised with Flat, and keep the CBOR
+format otherwise, we have defined the `serialiseUPLC` and `deserialiseUPLC` functions.
+
+Because Flat is not self-describing and it gets used in the encoding of Programs,
+data structures that include scripts (for example, transactions) no-longer benefit
+for CBOR's ability to self-describe it's format.
+-}
+
+serialiseUPLC :: UPLC.Program UPLC.DeBruijn DefaultUni DefaultFun () -> SerialisedScript
+serialiseUPLC =
+    -- See Note [Using Flat for serialsing/deserialising Script]
+    -- Currently, this is off because the old implementation didn't actually work, so we need to be careful
+    -- about introducing a working version
+    toShort . BSL.toStrict . serialise . SerialiseViaFlat
+
+deserialiseUPLC :: SerialisedScript -> UPLC.Program UPLC.DeBruijn DefaultUni DefaultFun ()
+deserialiseUPLC = unSerialiseViaFlat . deserialise . BSL.fromStrict . fromShort
+  where
+    unSerialiseViaFlat (SerialiseViaFlat a) = a
+
+serialiseScript :: forall a. CompiledCode a -> SerialisedScript
+serialiseScript = serialiseUPLC . toNameless . getPlc
+    where
+        toNameless :: UPLC.Program UPLC.NamedDeBruijn DefaultUni DefaultFun ()
```

Comment:
the `toNameless` traversal could be optimized away either by

1) coercing the program to Program FakeNamedDeBruijn which would break though the invariant that name=fakeName
2) write a newtype wrapper ThrowAwayNamedDeBrujin with deriving Flat via FakeNameDeBruijn and coerce to that

but maybe all these are not worth it because `serialiseScript` is run offline

Reply:
I see, so basically fusing the traversal with the flat serialisation.

It's a bit complicated, and this is not performance critical code as you said, so I'd leave it as it is.

Reply:
Please export this function.